search/includes/classes/class-queue.php

```diff
@@ -0,0 +1,350 @@
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->query( "TRUNCATE TABLE {$table_name}" ); // Cannot prepare table name. @codingStandardsIgnoreLine
+	}
+
+	public function count_jobs( $status, $object_type = 'post' ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT COUNT(*) FROM {$table_name} WHERE `status` = %s AND `object_type` = %s", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$status,
+				$object_type
+			)
+		);
+	}
+
+	public function count_jobs_due_now( $object_type = 'post' ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT COUNT(*) FROM {$table_name} WHERE `status` = 'queued' AND `object_type` = %s AND ( `start_time` <= NOW() OR `start_time` IS NULL )", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$object_type
+			)
+		);
+	}
+
+	public function get_next_job_for_object( $object_id, $object_type ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		$job = $wpdb->get_row(
+			$wpdb->prepare(
+				"SELECT * FROM {$table_name} WHERE `object_id` = %d AND `object_type` = %s AND `status` = 'queued' LIMIT 1", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$object_id,
+				$object_type
+			)
+		);
+
+		return $job;
+	}
+
+	public function get_batch_jobs( $count = 250 ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		// TODO transaction
+
+		$jobs = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT * FROM {$table_name} WHERE ( `start_time` <= NOW() OR `start_time` IS NULL ) AND `status` = 'queued' LIMIT %d", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$count
+			)
+		);
+
+		// Set them as running
+		$job_ids = wp_list_pluck( $jobs, 'id' );
+
+		$this->update_jobs( $job_ids, array( 'status' => 'running' ) );
+
+		// Set right status on the already queried jobs objects
+		foreach ( $jobs as &$job ) {
+			$job->status = 'running';
+		}
+
+		return $jobs;
+	}
+
+	public function process_batch_jobs( $jobs ) {
+		$indexables = \ElasticPress\Indexables::factory();
+	
+		// Organize by object type
+		$jobs_by_type = array();
+
+		foreach ( $jobs as $job ) {
+			if ( ! is_array( $jobs_by_type[ $job->object_type ] ) ) {
+				$jobs_by_type[ $job->object_type ] = array();
+			}
+
+			$jobs_by_type[ $job->object_type ][] = $job;
+		}
+		
+		// Batch process each type using the indexable
+		foreach ( $jobs_by_type as $type => $jobs ) {
+			$indexable = $indexables->get( $type );
+
+			$ids = wp_list_pluck( $jobs, 'object_id' );
+
+			$indexable->bulk_index( $ids );
+
+			// TODO handle errors
+
+			// Mark all as being indexed just now, for rate limiting
+			foreach ( $jobs as $job ) {
+				$this->set_last_index_time( $job->object_id, $job->object_type, time() );
+			}
+	
+			// Mark them as done in queue
+			$this->delete_jobs( $jobs );
```

Comment:
Rather than deleting them, should we mark them as `completed` and then have another process that handles the purging? Slightly higher cost, but would help with debugging (and could potentially be another signal to restrict high volume indexing for posts).

Reply:
Ideally yes, but the downside is we can no longer use our UNIQUE index on `object_id`, `object_type`, `status`. That index allows us to attempt to simply insert rows without having to first query to see if an object is queued, which is an important performance optimization when dealing with large post volumes.

If we mark as `complete`, we won't be able to mark any other jobs as `complete` until those rows are deleted, or we switch to first querying if an object is queued before attempting to insert the row (and drop the UNIQUE index).

Reply:
That makes sense :+1: